ykrt/src/location.rs

```diff
@@ -97,88 +29,407 @@ impl State {
+        let ls = self.load(Ordering::Relaxed);
+        debug_assert!(ls.is_locked());
+        debug_assert!(!ls.is_counting());
+        if self
+            .state
+            .compare_exchange(
+                ls.with_unparked().x,
+                ls.with_unparked().with_unlock().x,
+                Ordering::Release,
+                Ordering::Relaxed,
+            )
+            .is_ok()
+        {
+            return;
+        }
+
+        // At this point we know another thread has parked itself.
+        let key = unsafe { ls.hot_location() } as *const _ as usize;
+        debug_assert_ne!(key, 0);
+        let callback = |result: UnparkResult| {
+            if result.unparked_threads != 0 && result.be_fair {
+                if !result.have_more_threads {
+                    debug_assert!(ls.is_locked());
+                    self.state.store(ls.with_unparked().x, Ordering::Relaxed);
+                }
+                return TOKEN_HANDOFF;
+            }
+
+            if result.have_more_threads {
+                self.state
+                    .store(ls.with_unlock().with_parked().x, Ordering::Release);
+            } else {
+                self.state
+                    .store(ls.with_unparked().with_unlock().x, Ordering::Release);
+            }
+            TOKEN_NORMAL
+        };
+        unsafe {
+            unpark_one(key, callback);
+        }
     }
 
-    pub(crate) fn store(&self, state: State, order: Ordering) {
-        self.state.store(state.0, order);
+    /// Try obtaining the lock, returning the new `State` if successful.
+    pub(crate) fn try_lock(&self) -> Result<State<I>, ()> {
+        let mut ls = self.load(Ordering::Relaxed);
+        loop {
+            if ls.is_locked() {
+                return Err(());
+            }
+            let new_ls = ls.with_lock();
+            match self.compare_exchange_weak(ls, new_ls, Ordering::Acquire, Ordering::Relaxed) {
+                Ok(_) => return Ok(new_ls),
+                Err(x) => ls = x,
+            }
+        }
     }
 }
 
 impl<I> Drop for Location<I> {
     fn drop(&mut self) {
-        let lp = *self.state.get_mut();
-        if lp & PHASE_TAG == PHASE_TRACING {
-            unsafe { Arc::from_raw((lp & !PHASE_TAG) as *mut u8) };
+        let ls = self.load(Ordering::Relaxed);
+        if !ls.is_counting() {
+            debug_assert!(!ls.is_locked());
+            let hr = unsafe { ls.hot_location() };
+            unsafe {
+                Box::from_raw(hr);
+            }
+        }
+    }
+}
+
+#[cfg(target_pointer_width = "64")]
+pub(crate) const STATE_TAG: usize = 0b111; // All of the other tag data must fit in this.
+#[cfg(target_pointer_width = "64")]
+pub(crate) const STATE_NUM_BITS: usize = 3;
+
+pub(crate) const STATE_IS_LOCKED: usize = 0b001;
+pub(crate) const STATE_IS_PARKED: usize = 0b010;
+pub(crate) const STATE_IS_COUNTING: usize = 0b100;
+
+pub(crate) const TOKEN_NORMAL: UnparkToken = UnparkToken(0);
```

Comment:
Do these make more sense as an enum with explicit discriminant numbers? I appreciate that the example code does the same... up to you.

Reply:
These are from `parking_lot_core` so we can't encode them as an `enum`.

Reply:
Ah, ok.